src/persist/src/file.rs

```diff
@@ -182,6 +183,198 @@ impl Persister for FilePersister {
+        base_dir.join(Self::BUFFER_PATH)
+    }
+
+    fn metadata_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::METADATA_PATH)
+    }
+}
+
+impl Buffer for FileBuffer {
+    fn write_sync(&mut self, buf: Vec<u8>) -> Result<SeqNo, Error> {
+        self.buf.clear();
+        let write_seqno = self.seqno.end;
+        self.seqno = self.seqno.start..SeqNo(self.seqno.end.0 + 1);
+        // Write length prefixed data, and then the sequence number.
+        let len = buf.len() as u64;
+
+        self.buf.extend(&len.to_le_bytes());
+        self.buf.extend(buf);
+        self.buf.extend(&write_seqno.0.to_le_bytes());
+        self.buf.shrink_to_fit();
+
+        let guard = self.dataz.lock()?;
+        let mut buffer_file = &guard.0;
+        buffer_file.write_all(&self.buf)?;
+        buffer_file.sync_all()?;
+        Ok(write_seqno)
+    }
+
+    fn snapshot<F>(&self, mut logic: F) -> Result<Range<SeqNo>, Error>
+    where
+        F: FnMut(SeqNo, &[u8]) -> Result<(), Error>,
+    {
+        let mut bytes = Vec::new();
+
+        {
+            let guard = self.dataz.lock()?;
+            let mut buffer_file = &guard.0;
+            buffer_file.seek(SeekFrom::Start(0))?;
+            buffer_file.read_to_end(&mut bytes)?;
+        }
+
+        let mut bytes = &bytes[..];
+        let mut buf = [0u8; 8];
+        while !bytes.is_empty() {
+            // Decode the data
+            bytes.read_exact(&mut buf)?;
+            let data_len = u64::from_le_bytes(buf);
+            let mut data = vec![0u8; data_len.try_into().unwrap()];
+            bytes.read_exact(&mut data[..])?;
+
+            // Decode sequence number
+            bytes.read_exact(&mut buf)?;
+            let seqno = SeqNo(u64::from_le_bytes(buf));
+
+            if seqno < self.seqno.start {
+                // This record has already been truncated so we can ignore it.
+                continue;
+            } else if seqno >= self.seqno.end {
+                return Err(format!(
+                    "invalid sequence number {:?} found for buffer containing {:?}",
+                    seqno, self.seqno
+                )
+                .into());
+            }
+
+            logic(seqno, &data)?;
+        }
+        Ok(self.seqno.clone())
+    }
+
+    /// Logically truncates the buffer so that future reads ignore writes at sequence numbers
+    /// less than `upper`.
+    ///
+    /// TODO: actually reclaim disk space as part of truncating.
+    fn truncate(&mut self, upper: SeqNo) -> Result<(), Error> {
+        // TODO: Test the edge cases here.
+        if upper <= self.seqno.start || upper > self.seqno.end {
+            return Err(format!(
+                "invalid truncation {:?} for buffer containing: {:?}",
+                upper, self.seqno
+            )
+            .into());
+        }
+        self.seqno = upper..self.seqno.end;
+
+        self.buf.clear();
+        self.buf.extend(&self.seqno.end.0.to_le_bytes());
```

Comment:
nit: i think you can just `let len_raw = self.seqno.end.0.to_le_bytes()` here since it essentially just transmutes and this all stays on the stack

Reply:
ah whoops this should be seqno.start :/